src/checkers/inference/solver/backend/z3smt/Z3SmtSolver.java

```diff
@@ -0,0 +1,429 @@
+    protected static final String constraintsFile = pathToProject + "/z3Constraints.smt";
+    protected static final String constraintsUnsatCoreFile =
+            pathToProject + "/z3ConstraintsUnsatCore.smt";
+    protected static final String constraintsStatsFile = pathToProject + "/z3ConstraintsGlob.smt";
+
+    // timing statistics variables
+    protected long serializationStart;
+    protected long serializationEnd;
+    protected long solvingStart;
+    protected long solvingEnd;
+
+    /** Whether the timeout arg should be set */
+    protected boolean withTimeout = false;
+
+    public Z3SmtSolver(
+            SolverEnvironment solverEnvironment,
+            Collection<Slot> slots,
+            Collection<Constraint> constraints,
+            Z3SmtFormatTranslator<SlotEncodingT, SlotSolutionT> z3SmtFormatTranslator,
+            Lattice lattice) {
+        super(solverEnvironment, slots, constraints, z3SmtFormatTranslator, lattice);
+
+        Map<String, String> z3Args = new HashMap<>();
+
+        if (withTimeout) {
+            z3Args.put("timeout", Integer.toString(timeout()));
+        }
+        // creating solver
+	    ctx = new Context(z3Args);
+
+        z3SmtFormatTranslator.init(ctx);
+    }
+
+    protected int timeout() {
+        return 2 * 60 * 1000; // timeout of 2 mins by default
+    }
+
+    // Main entry point
+    @Override
+    public Map<Integer, AnnotationMirror> solve() {
+        // serialize based on user choice of running in optimizing or non-optimizing mode
+        optimizingMode = solverEnvironment.getBoolArg(Z3SolverEngineArg.optimizingMode);
+        getUnsatCore = false;
+
+        if (optimizingMode) {
+            logger.fine("Encoding for optimizing mode");
+        } else {
+            logger.fine("Encoding for non-optimizing mode");
+        }
+
+        serializeSMTFileContents();
+
+        solvingStart = System.currentTimeMillis();
+        // in Units, if the status is SAT then there must be output in the model
+        List<String> results = runZ3Solver();
+        solvingEnd = System.currentTimeMillis();
+
+        Statistics.addOrIncrementEntry(
+                "smt_serialization_time(millisec)", serializationEnd - serializationStart);
+        Statistics.addOrIncrementEntry("smt_solving_time(millisec)", solvingEnd - solvingStart);
+
+        if (results == null) {
+            logger.fine("!!! The set of constraints is unsatisfiable! !!!");
+            return null;
+        }
+        
+        return formatTranslator.decodeSolution(
+                        results, solverEnvironment.processingEnvironment);
+    }
+
+    @Override
+    public Collection<Constraint> explainUnsatisfiable() {
+        optimizingMode = false;
+        getUnsatCore = true;
+
+        logger.fine("Now encoding for unsat core dump.");
+        serializeSMTFileContents();
+
+        solvingStart = System.currentTimeMillis();
+        runZ3Solver();
```

Comment:
Add a comment here why the result of the call is ignored - as it's unsat, it's always null?

Reply:
This seems to be the second run if it's UNSAT. Why run the solver a second time instead of making use of the result from the first run?